### PR TITLE
refactor(migrations): migrate regexManagers

### LINKED ISSUE #19066: Rename regexManagers to customManagers
### Describe the proposed change(s).

- rename `regexManagers` to `customManagers`
- add a new field `customType` which defaults to `regex`
- add migration code

### PR DESCRIPTION
**Pull request by RahulGautamSingh**
## Changes
- migrate `regexManagers` to `customManagers`
- rename the manager from `regex` to `custom`
- add new wrapper for custom managers

## Context
- Closes: https://github.com/renovatebot/renovate/issues/19066

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/config/presets/internal/index.ts [hunk 1]
```diff
@@ -9,7 +9,7 @@ import * as monorepoPreset from './monorepo';
 import * as npm from './npm';
 import * as packagesPreset from './packages';
 import * as previewPreset from './preview';
-import * as regexManagersPreset from './regex-managers';
+import * as customManagersPreset from './regex-managers';
```
Comment (viceice) [inline lib/config/presets/internal/index.ts:12]: we should also rename that file

File: lib/config/presets/internal/regex-managers.ts [hunk 2]
```diff
@@ -4,14 +4,14 @@ import type { Preset } from '../types';
 
 export const presets: Record<string, Preset> = {
   dockerfileVersions: {
-    description: 'Update `_VERSION` variables in Dockerfiles.',
```
Comment (viceice) [inline lib/config/presets/internal/regex-managers.ts:7]: why do you moved the description, please move back to original place

File: lib/config/types.ts [hunk 3]
```diff
@@ -179,6 +179,7 @@ export interface RegexManagerTemplates {
   registryUrlTemplate?: string;
 }
 export interface RegExManager extends RegexManagerTemplates {
```
Comment (viceice) [inline lib/config/types.ts:181]: interfaces needs to be renamed too

File: lib/config/validation.spec.ts [hunk 4]
```diff
@@ -309,69 +309,70 @@ describe('config/validation', () => {
...
+            },
+          ],
+        };
+        const { warnings, errors } = await configValidation.validateConfig(
+          config,
+          true
+        );
+        expect(warnings).toHaveLength(0);
+        expect(errors).toHaveLength(1);
+        expect(errors).toMatchSnapshot();
```
Comment (viceice) [inline lib/config/validation.spec.ts:330]: can we use toMatchObject instead please

File: lib/config/presets/internal/index.ts (hunk 1 above)
Comment (rarkins) [reply lib/config/presets/internal/index.ts:12]: Also needs a migration

File: lib/modules/manager/custom/index.ts [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+import managers from './api';
+
+export function extractPackageFile(
+  content: string,
+  fileName: string,
+  config: ExtractConfig
+): Result<PackageFile | null> {
+  const { customType = 'regex' } = config;
+  if (!managers.has(customType)) {
+    return null;
```
Comment (rarkins) [inline lib/modules/manager/custom/index.ts:11]:
```suggestion
    logger.warn({ customType }, 'Unknown customManagers type');
    return null;
```

File: lib/modules/manager/custom/readme.md [hunk 6]
```diff
@@ -0,0 +1 @@
+This is a custom wrapper for the current regex manager and any future custom implmentations to come.
```
Comment (rarkins) [inline lib/modules/manager/custom/readme.md:1]:
```suggestion
This is a custom wrapper for the current regex manager and any future custom implementations to come.
```

File: lib/modules/manager/custom/regex/readme.md [hunk 7]
```diff
@@ -4,17 +4,17 @@ The `regex` manager is unique in Renovate because:
```
Comment (rarkins) [inline lib/modules/manager/custom/regex/readme.md:4]: Change the `regex` mentions above this but also add a note that "customManagers was formerly known as regexManagers" for anyone searching the docs

File: lib/modules/manager/index.ts [hunk 8]
```diff
@@ -65,6 +66,11 @@ export function extractPackageFile(
   if (!managers.has(manager)) {
     return null;
   }
+
+  if (manager === 'custom') {
+    return customManager.extractPackageFile(content, fileName, config);
+  }
```
Comment (rarkins) [inline lib/modules/manager/index.ts:72]: Why is this change needed?

File: lib/workers/types.ts [hunk 9]
```diff
@@ -151,13 +151,13 @@ export interface BranchSummary {
 }
 
 export interface WorkerExtractConfig extends ExtractConfig {
+  customManagers?: RegExManager[];
```
Comment (rarkins) [inline lib/workers/types.ts:154]:
```suggestion
  customManagers?: CustomManager[];
```

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/config/presets/internal/regex-managers.ts (hunk 2 above)
Comment (RahulGautamSingh) [reply lib/config/presets/internal/regex-managers.ts:7]: Cz of the sort-keys comment above.

File: lib/modules/manager/index.ts (hunk 8 above)
Comment (RahulGautamSingh) [reply lib/modules/manager/index.ts]: Not needed reverting.

File: lib/config/validation.spec.ts [hunk 10]
```diff
@@ -288,7 +288,7 @@ describe('config/validation', () => {
         true
       );
       expect(warnings).toHaveLength(0);
-      expect(errors).toMatchSnapshot();
+      expect(errors).toMatchObject([]);
```
Comment (viceice) [inline lib/config/validation.spec.ts:291]: use to be empty array and remove check below use same for warnings above

File: lib/config/validation.spec.ts [hunk 11]
```diff
@@ -309,178 +309,191 @@ describe('config/validation', () => {
...
+              datasourceTemplate: 'maven',
+              versioningTemplate: 'gradle',
+            },
+          ],
+        };
+        const { warnings, errors } = await configValidation.validateConfig(
+          config,
+          true
+        );
+        expect(warnings).toHaveLength(0);
```
Comment (viceice) [inline lib/config/validation.spec.ts:328]: see other comment

File: lib/config/validation.spec.ts [hunk 12]
```diff
@@ -309,178 +309,191 @@ describe('config/validation', () => {
...
+        expect(errors).toHaveLength(1);
```
Comment (viceice) [inline lib/config/validation.spec.ts:329]: can be removed, check below enforce array length match

File: lib/config/validation.spec.ts [hunk 13]
```diff
@@ -309,178 +309,191 @@ describe('config/validation', () => {
...
+        expect(errors).toMatchObject([
```
Comment (viceice) [inline lib/config/validation.spec.ts:330]: does toEqual work too or do we have more properties to ignore for the test?

File: lib/modules/manager/custom/index.ts [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+  const { customType = 'regex' } = config;
+  if (!managers.has(customType)) {
+    logger.warn({ customType }, 'Unknown customManagers type');
+    return null;
+  }
+
+  const m = managers.get(customType)!;
+  return m?.extractPackageFile
+    ? m.extractPackageFile(content, fileName, config)
+    : null;
```
Comment (viceice) [inline lib/modules/manager/custom/index.ts:19]:
```suggestion
  return m.extractPackageFile!.(content, fileName, config);
```
we should use a custom manager  API interface and make that function mandatory.

File: lib/config/validation.spec.ts (hunk 13 above)
Comment (RahulGautamSingh) [reply lib/config/validation.spec.ts:330]: `toEqual` doesn't care about array's length, whereas `toMatchObject` does. So using it and removing the 
`toHaveLength` assestions.

Comment (RahulGautamSingh) [reply lib/config/validation.spec.ts:330]: it works

File: lib/modules/manager/custom/regex/readme.md [hunk 15]
```diff
@@ -125,53 +127,8 @@ But we included the `versioningTemplate` config option to show you why we call t
...
 This is because Handlebars escapes special characters with double braces (by default).
 
-By adding `renovate: datasource=` and `depName=` comments to the `Dockerfile` you only need _one_ `regexManager` instead of _four_.
+By adding `renovate: datasource=` and `depName=` comments to the `Dockerfile` you only need _one_ `customManager` instead of _four_.
 The `Dockerfile` is documented better as well.
 
 The syntax in the example is arbitrary and you can set your own syntax.
 If you do, update your `matchStrings` regex!
-
-### Using regexManager to update the dependency name in addition to version
```
[the remark this answers is not in the document]
Comment (RahulGautamSingh) [reply lib/modules/manager/custom/regex/readme.md:134]: not intentional, typo maybe...reverting.

File: lib/modules/manager/custom/regex/readme.md [hunk 16]
```diff
@@ -1,20 +1,22 @@
-With the `regex` manager you can configure Renovate so it finds dependencies that are not detected by its other built-in package managers.
+With the `custom` manager you can configure Renovate so it finds dependencies that are not detected by its other built-in package managers.
 
-The `regex` manager is unique in Renovate because:
+**Note:** `customManagers` was formerly known as `regexManagers`
```
[the remark this answers is not in the document]
Comment (RahulGautamSingh) [reply lib/modules/manager/custom/regex/readme.md:3]: Working on a new defn for custom Managers 🤞

**Review by viceice — COMMENTED** [review]
needs extensive real run tests

File: lib/config/presets/internal/custom-managers.ts [hunk 17]
```diff
@@ -4,8 +4,7 @@ import type { Preset } from '../types';
 
 export const presets: Record<string, Preset> = {
   dockerfileVersions: {
-    description: 'Update `_VERSION` variables in Dockerfiles.',
```
Comment (rarkins) [inline lib/config/presets/internal/custom-managers.ts:7]: This seems like it wasn't copy pasted new?

File: lib/config/types.ts [hunk 18]
```diff
@@ -382,7 +383,7 @@ export interface RenovateArrayOption<
 }
 
 export interface RenovateStringArrayOption extends RenovateArrayOption<string> {
-  format?: 'regex';
+  format?: 'CustomManagerTemplates';
```
Comment (rarkins) [inline lib/config/types.ts:386]: I'm a bit confused by this part

File: lib/config/presets/internal/custom-managers.ts (hunk 17 above)
Comment (RahulGautamSingh) [reply lib/config/presets/internal/custom-managers.ts:7]: Still working on a bad merge.

**Comment by rarkins** [comment]
@RahulGautamSingh it might be easier to restart/reapply these changes than to try to resolve the conflicts?
